torch/csrc/jit/passes/peephole.cpp

```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
           u.user->replaceInput(0, node->inputs().at(0));
         }
       }
+    } else if (node->kind() == prim::If) {
+      IfView n(node);
+      // this handles redundant short circuits like "x and True" or "x or False"
+      for (size_t i = 0; i < n.outputs().size(); ++i) {
+        if (n.outputs().at(i)->type() != BoolType::get()) {
+          continue;
+        }
+        bool true_val =
+            constant_as<bool>(n.thenOutputs().at(i)).value_or(false);
+        bool false_val =
+            constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+        // if an if node's output equals its condition replace output with
+        // condition
+        if (true_val && !false_val) {
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
+        }
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
+        bool input_must_be_none =
+            node->inputs().at(check_none_index)->mustBeNone();
+        Value* other_input = node->inputs().at(1 - check_none_index);
+        bool other_must_not_be_none = !other_input->mustBeNone() &&
+            !other_input->type()->cast<OptionalType>();
+        if (input_must_be_none && other_must_not_be_none) {
+          WithInsertPoint guard(node);
+          auto output = node->owningGraph()->insertConstant(
+              node->kind() == aten::__isnot__);
+          node->output()->replaceAllUsesWith(output);
+        }
+      }
+    } else if (
+        node->kind() == prim::unchecked_unwrap_optional ||
+        node->kind() == aten::_unwrap_optional) {
+      // we are unwrapping an input that can't be None, remove the unwrap
+      auto input = node->input();
+      if (input->type() != NoneType::get() &&
+          !input->type()->cast<OptionalType>()) {
```

Comment:
Should this also cover AutogradZeroTensor?

Reply:
I talked with @wanchaol on Friday about this and IIRC he seemed to think mustBeNone doesn't need to check for AutogradZero. @wanchaol  could you weigh in?

Reply:
As far as I could tell, AutogradZero is undefined tensors fed into `Optional[Tensor]` inputs for backward graphs.
I thought I saw a AutogradZero being treated as None somewhere, but I don't find it at the moment.